src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+    if (ret < 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::failed rgw_rados_notify("
+			<< DEDUP_WATCH_OBJ << ")::err="<<cpp_strerror(ret) << dendl;
+      return ret;
+    }
+    std::vector<librados::notify_ack_t> acks;
+    std::vector<librados::notify_timeout_t> timeouts;
+    ioctx.decode_notify_response(reply_bl, &acks, &timeouts);
+    if (timeouts.size() > 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::failed rgw_rados_notify("
+			<< DEDUP_WATCH_OBJ << ")::timeout error" << dendl;
+      return -EAGAIN;
+    }
+
+    for (auto& ack : acks) {
+      try {
+	ldpp_dout(dpp, 20) << __func__ << "::ACK: notifier_id=" << ack.notifier_id
+			   << "::cookie=" << ack.cookie << dendl;
+	auto iter = ack.payload_bl.cbegin();
+	ceph::decode(ret, iter);
+	struct rgw::dedup::control_t ctl;
+	decode(ctl, iter);
+	ldpp_dout(dpp, 10) << __func__ << "::++ACK::ctl=" << ctl << "::ret=" << ret << dendl;
+      } catch (buffer::error& err) {
+	ldpp_dout(dpp, 1) << __func__ << "::failed decoding notify acks" << dendl;
+	return -EINVAL;
+      }
+      if (ret != 0) {
+	ldpp_dout(dpp, 1) << __func__ << "::Bad notify ack, ret=" << ret
+			  << "::err=" << cpp_strerror(ret) << dendl;
+	return ret;
+      }
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::" << get_urgent_msg_names(urgent_msg)
+		       << " finished successfully!" << dendl;
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  // command-line called from radosgw-admin.cc
+  int cluster::dedup_restart_scan(rgw::sal::RadosStore *store,
+				  bool dry_run,
+				  const DoutPrefixProvider *dpp)
+  {
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      return -1;
+    }
+
+    dedup_epoch_t old_epoch;
+    // store the previous epoch for cmp-swap
+    int ret = get_epoch(ioctx, dpp, &old_epoch, __func__);
+    if (ret != 0) {
+      return ret;
+    }
+
+    // first abort all dedup work!
+    ret = dedup_control(store, dpp, URGENT_MSG_ABORT);
+    if (ret != 0) {
+      return ret;
+    }
+
+    ldpp_dout(dpp, 1) << __func__ << (dry_run ?"::DRY RUN":"::FULL DEDUP") << dendl;
+    int dedup_type = (dry_run ? DEDUP_TYPE_DRY_RUN : DEDUP_TYPE_FULL);
+    ret = swap_epoch(dpp, ioctx, &old_epoch, dedup_type, 0, 0);
+    if (ret == 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::Epoch object was reset" << dendl;
+      dedup_control(store, dpp, URGENT_MSG_RESTART);
+    }
+    else {
+      ldpp_dout(dpp, 0) << __func__ << "::Failed to reset EPOCH object with: "
+			<< cpp_strerror(ret) << ", ret=" << ret << dendl;
+    }
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  bool cluster::can_start_new_scan(rgw::sal::RadosStore *store)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::epoch=" << d_epoch_time << dendl;
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      return -1;
```

Comment:
```suggestion
      return false;
```

Reply:
change the code to return the system error code reported by init_dedup_pool_ioctx()